src/TypeScript/EslintBridgeClient/EslintBridgeKeepAlive.cs

```diff
@@ -0,0 +1,110 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.SystemAbstractions;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.Integration.Helpers;
+
+namespace SonarLint.VisualStudio.TypeScript.EslintBridgeClient
+{
+    /// <summary>
+    /// Component that is responsible for sending keep-alive messages
+    /// to the eslint-bridge
+    /// </summary>
+    internal interface IEslintBridgeKeepAlive : IDisposable
+    {
+    }
+
+    internal sealed class EslintBridgeKeepAlive : IEslintBridgeKeepAlive
+    {
+        // The eslintbridge process will shutdown after 15 seconds.
+        // We send a keep alive every 5 seconds.
+        private const double MillisecondsToWaitBetweenKeepAlives = 5000;
+
+        private readonly IEslintBridgeProcess process;
+        private readonly IEslintBridgeHttpWrapper httpWrapper;
+        private readonly ITimer timer;
+        private readonly ILogger logger;
+
+        public EslintBridgeKeepAlive(IEslintBridgeProcess process, ILogger logger)
+            : this(process, logger, new EslintBridgeHttpWrapper(logger), new TimerWrapper())
+        {
+        }
+
+        internal /* for testing */ EslintBridgeKeepAlive(IEslintBridgeProcess process, ILogger logger,
+            IEslintBridgeHttpWrapper httpWrapper, ITimer timer)
```

Comment:
I feel that the dependencies between this class and EslintBridgeClient show a problem.

EslintBridgeClient is supposed to be the only one that knows how to communicate with the eslint bridge process; it's responsible for ensuring the process is up and building the server URL. It's a bit out of place that EslintBridgeClient is the one in charge of the timer.

How about a different approach:
* `EslintBridgeAnalyzer` can be the one to create and dispose the `EslintBridgeKeepAlive` class.
* `EslintBridgeKeepAlive` will only have the timer, and will call `EslintBridgeClient.SendKeepAlive()`.
* `EslintBridgeClient` then becomes simple and only does the requests, like it did before.

WDYT?

Reply:
> It's a bit out of place that EslintBridgeClient is the one in charge of creating and disposing the timer class and that this class also knows how to build the URL and get the port.

I agree with the second bit, but I don't think it is out of place that the EsLintBridgeClient is ultimately responsible for starting the process and keeping it running.

The `EsLintBridgeClient` hides the fact that there is an external process, and handles recovery when the external process fails. I think the fact that the keep-alive is required should be hidden from callers too, including the analyzer.

 Instead of duplicating the URL building, one option is to move it to the http wrapper. The other would be to add a KeepAlive to the EsLintBridgeClient, and have the keep-alive timer class call back to that.

What do you think? (a quick call might be easier, if it's not too noisy).

Reply:
As discussed offline, we don't want to change the analyzer. I'll merge this as-is and look at changing the keep-alive class separately.